sdb/parser.py

```diff
@@ -0,0 +1,206 @@
+
+    Example:
+
+        sdb> cmd0 arg0 | cmd1 arg1 "arg 2" ! shell_cmd shell_args ...
+             ---------   -----------------   ========================
+
+    Everything underlined with '-' is a CMD (e.g. SDB command) and
+    '=' is a SHELL_CMD token.
+    """
+    CMD = "__cmd__"
+    SHELL_CMD = "__shell_cmd__"
+
+
+WHITESPACE = " \t"
+QUOTES = '"\''
+OPERATORS = "|!"
+DELIMETERS = OPERATORS + QUOTES + WHITESPACE
+
+
+def _next_non_whitespace(line: str, index: int) -> Optional[int]:
+    """
+    Return the index of the next non-whitespace character in `line`
+    starting from `index` or None if there is no such character until
+    the end of `line`.
+    """
+    for i, c in enumerate(line[index:]):
+        if c not in WHITESPACE:
+            return i + index
+    return None
+
+
+def _next_delimiter(line: str, index: int) -> Optional[int]:
+    """
+    Return the index of the next delimeter in `line` starting from
+    `index` or None if there is no such character until the end of
+    `line`. Generally used when we are in the middle of processing
+    an identifier/token and want to see where it ends.
+    """
+    for i, c in enumerate(line[index:]):
+        if c in DELIMETERS:
+            return i + index
+    return None
+
+
+def tokenize(line: str) -> Iterable[Tuple[List[str], ExpressionType]]:
+    """
+    Iterates over the line passed as an input (usually from the REPL) and
+    generates expressions to be evaluated by the SDB pipeline logic. The
+    actual expression information vary by expression type:
+
+    [1] CMD (e.g. SDB commands) expression contain a list of strings
+        which contains the command (first string of list) and its
+        arguments (the rest of the strings in the list).
+    [2] A SHELL_CMD expression (e.g. basically anything after a bang !)
+        is a single string that contains the whole shell command,
+        including its arguments and the spaces between them.
+
+    Example:
+
+        sdb> cmd0 arg0 | cmd1 arg1 "arg 2" ! shell_cmd shell_args ...
+             ---------   -----------------   ========================
+
+        Returns:
+            Iterable [
+                 (['cmd0', 'arg0'], CMD),
+                 (['cmd1', 'arg1', 'arg 2'], CMD),
+                 (['shell_cmd shell_args ...'], SHELL_CMD),
+            ]
+
+    Note: The reason that we split the arguments for CMDs here is so we
+    don't have to redo that work later in the Command class where we need
+    to parse the arguments in argparse. Furthermore, the tokenizer here
+    does a better job than doing a simple split() as it parses each string
+    containing spaces as a single argument (e.g. space in "arg 2" in our
+    example).
+    """
+    # pylint: disable=too-many-statements
+    # pylint: disable=too-many-branches
+
+    token_list: List[str] = []
+    idx: Optional[int] = 0
+    while True:
+        idx = _next_non_whitespace(line, idx)  # type: ignore[arg-type]
```

Comment:
Not a nit, just kinda surprised mypy doesn't realize that idx is never NONE here